### ISSUE TITLE
Use Flux to calculate deadman alerts

### ISSUE THREAD

**Issue by nisley** [influxdata/flux#1338]
The alerts & notifications design has deadman alert types as a requirement. A deadman alert is an alert that data has stopped reporting for some period of time.

Currently, we believe Flux cannot be used to calculate if a deadman condition has been reached.

**Comment by nathanielc** [comment]
The limitation that we face for computing a deadman is that the InfluxDB source cannot produce 0 count results for series that have not written data. 

A naive deadman would look like this:

```
from(bucket:"foo")
    |> range(start: -1m)
    |> filter(fn: (r) => r._measurement == "process")
    |> count()
    |> alert(
         crit: (r) => r.count == 0 
    ) // made up syntax for defining an alert when the count goes to 0
```

The issue is that given a series `process=bar` is reporting data and then stop reporting data, then the query to count the points for the 1m time range will not return any value of the series `process=bar` because it doesn't have any data during the interval. 

So relying on the InfluxDB source to produce 0 counts for dead series is not a practical solution to a deadman alert.

We need to explore possible solutions to this issue. A few past suggestions that have come up:

1. Use the series index to produce 0 counts. This could work but would result in unexpected behavior as the end user doesn't have direct control over which series remain in the series index and for how long. 
2. Do some kind of reverse walk through the tsm data to find old series that may have been reporting but are no longer reporting. This has lots of questions: How far back do we look? How does the user specify how far back to look? Is this practical from a performance perspective?
3. Create and use some kind of time based index that can be consulted. This is similar to option (1) but different in that the index can be controlled in the time dimension allowing for a more clear understanding of what is in the index.

All of these solutions leave open the question of how far back in time do we search for series that we would expect to keep reporting but haven't.

A less naive deadman query would be something like this:

```
recent = from(bucket:"foo")
    |> range(start: -1m)
    |> filter(fn: (r) => r._measurement == "process")
    |> count()

previous = 
from(bucket:"foo")
    |> range(start: -1h)
    |> filter(fn: (r) => r._measurement == "process")
    |> count()

// Perform outer join between recent and previous data.
join(tables:{recent, previous}, on:["_stop", ...], mode: "outer", )
    |> alert(
         crit: (r) => isNull(v: r.count_recent)
    ) // made up syntax for defining an alert when the recent count is null i.e is was not existant
```

Flux cannot perform outer joins currently. If this is the solution we go with then we need to create a helper deadman method to encapsulate this logic is I would not expect a user to be able to easily construct such a query.

**Comment by nathanielc** [comment]
DOD:

* We need a design for how to perform deadman alerts from Flux
* Create issues for implementing the design

**Comment by aanthony1243** [comment]
thinking out loud, would:  changing nothing about flux and then calling `assertEmpty` on the result work?

**Comment by nathanielc** [comment]
@aanthony1243 No, as you still wouldn't know which series didn't return any data.

**Comment by jsternberg** [comment]
Something I would really like to see in the design is how we are going to do triggers because I think triggers play a very large role in this and I'd like to see them fleshed out.

As a quick example of what I'm thinking, Kapacitor will calculate a window after it receives a point that is part of the next group because it makes the assumption that no more points will be needed. This has its flaws, which should be covered, but it's an interesting idea and the idle barrier works really well! And when Kapacitor receives the points on time, this works exactly as you expect.

And I wonder if that could be represented by triggers and how that would be represented. I wonder if a trigger for the latest time could be sent after reading a series from the database and how that would be represented by the downstream transformations. If triggers aren't an effective method of implementing this behavior, I want to see that because otherwise I think triggers could be removed.

I'd really like to see that be part of whatever document or I would like to help develop those ideas so we could figure out how they work. I don't find the current usage and implementation of triggers to be very effective, but I think they lay a great foundation once they get fleshed out a little bit and the only purpose of triggers in the flux codebase is to lay out that foundation at the moment.

**Comment by nathanielc** [comment]
I also see triggers as only a foundational piece at the moment that aren't really leveraged. 

I don't see deadman as a _needing_ the streaming behavior that triggers enable. A deadman in its simplest form is just a periodic check for missing data that existed in a previous interval. Triggers are not really needed for that as we have tasks to schedule the periodic execution of the check.

I do see triggers as enabling more advanced deadman queries. For example a deadman query could be defined to evaluate on a trigger that fires with each new data point in addition to an idle timer. Such behavior would allow for reducing the latency of detecting dead data. But until we have the concept of a continuously running stream query we don't really have the ability to leverage triggers in that way. Supporting true streaming queries is not a near term objective.

**Comment by pauldix** [comment]
Note  that this design has been updated based on discussion here: https://github.com/influxdata/flux/pull/1588#pullrequestreview-268731198

**Comment by pauldix** [comment]
Closed by #1650
